synapse/rest/client/v1/login.py

```diff
@@ -254,6 +283,118 @@ def on_GET(self, request):
         return (200, {"serverUrl": self.cas_server_url})
 
 
+class CasRedirectServlet(ClientV1RestServlet):
+    PATTERN = client_path_pattern("/login/cas/redirect")
+
+    def __init__(self, hs):
+        super(CasRedirectServlet, self).__init__(hs)
+        self.cas_server_url = hs.config.cas_server_url
+        self.cas_service_url = hs.config.cas_service_url
+
+    def on_GET(self, request):
+        args = request.args
+        if "redirectUrl" not in args:
+            return (400, "Redirect URL not specified for CAS auth")
+        clientRedirectUrlParam = urllib.urlencode({
+            "redirectUrl": args["redirectUrl"][0]
+        })
+        hsRedirectUrl = self.cas_service_url + "/_matrix/client/api/v1/login/cas/ticket"
+        serviceParam = urllib.urlencode({
+            "service": "%s?%s" % (hsRedirectUrl, clientRedirectUrlParam)
+        })
+        request.redirect("%s?%s" % (self.cas_server_url, serviceParam))
+        request.finish()
+        defer.returnValue(None)
+
+
+class CasTicketServlet(ClientV1RestServlet):
+    PATTERN = client_path_pattern("/login/cas/ticket")
+
+    def __init__(self, hs):
+        super(CasTicketServlet, self).__init__(hs)
+        self.cas_server_url = hs.config.cas_server_url
+        self.cas_service_url = hs.config.cas_service_url
+        self.cas_required_attributes = hs.config.cas_required_attributes
+
+    @defer.inlineCallbacks
+    def on_GET(self, request):
+        clientRedirectUrl = request.args["redirectUrl"][0]
+        # TODO: get this from the homeserver rather than creating a new one for
```

Comment:
http_client = self.hs.get_simple_http_client()

(Sorry that this is so undiscoverable)